code/drasil-theory/Theory/Drasil/ModelKinds.hs

```diff
@@ -1,42 +1,63 @@
+{-# LANGUAGE TemplateHaskell, Rank2Types, ScopedTypeVariables, PostfixOperators  #-}
+module Theory.Drasil.ModelKinds (ModelKinds(..), RealmVariant(..),
+  setMk, elimMk, lensMk, getEqModQds) where
 
-{-# LANGUAGE TemplateHaskell, Rank2Types, ScopedTypeVariables  #-}
-module Theory.Drasil.ModelKinds (ModelKinds(..), setMk, elimMk, lensMk, getEqModQds) where
-
-import Language.Drasil (QDefinition, RelationConcept, ConceptDomain(..),
-                        Definition(..), ExprRelat(..), Idea(..), 
-                        NamedIdea(..), HasUID(..))
-import Control.Lens (Lens', (^.), to, lens, set, makeLenses, Getter, Setter')
+import Control.Lens ((^.), to, lens, set, makeLenses, makeLensesFor,
+  Getter, Lens', Setter')
 import Data.Maybe (mapMaybe)
+import qualified Data.List.NonEmpty as NE
+
+import Language.Drasil (($=), sy, Expr, RelationConcept,
+  NamedIdea(..), HasUID(..), ExprRelat(..), ConceptDomain(..), Definition(..),
+  Idea(..), DefiningExpr(..), UID, Sentence, QDefinition, QuantityDict)
+
+data RealmVariant = RV {
+  _desc :: Sentence,
+  _expr :: Expr
+}
+
+makeLensesFor [("_expr", "expr")] ''RealmVariant
 
 data ModelKinds = EquationalModel QDefinition
+                | EquationalRealm QuantityDict [UID] (NE.NonEmpty RealmVariant)
                 | DEModel RelationConcept
                 | OthModel RelationConcept
 
 makeLenses ''ModelKinds
 
-instance HasUID             ModelKinds where uid = lensMk uid uid
-instance NamedIdea          ModelKinds where term = lensMk term term
-instance Idea               ModelKinds where getA = elimMk (to getA) (to getA)
-instance Definition         ModelKinds where defn = lensMk defn defn
-instance ConceptDomain      ModelKinds where cdom = elimMk (to cdom) (to cdom)
-instance ExprRelat          ModelKinds where relat = elimMk (to relat) (to relat)
-
-elimMk :: Getter QDefinition a -> Getter RelationConcept a -> ModelKinds -> a
-elimMk l _ (EquationalModel q) = q ^. l
-elimMk _ l (DEModel q)         = q ^. l
-elimMk _ l (OthModel q)        = q ^. l
-
-setMk :: ModelKinds -> Setter' QDefinition a -> Setter' RelationConcept a -> a -> ModelKinds
-setMk (EquationalModel q) f _ x = EquationalModel $ set f x q
-setMk (DEModel q)         _ g x = DEModel $ set g x q
-setMk (OthModel q)        _ g x = OthModel $ set g x q
-
-lensMk :: forall a. Lens' QDefinition a -> Lens' RelationConcept a -> Lens' ModelKinds a
-lensMk lq lr = lens g s
+instance HasUID        ModelKinds where uid      = lensMk uid uid uid
+instance NamedIdea     ModelKinds where term     = lensMk term term term
+instance Idea          ModelKinds where getA     = elimMk (to getA) (to getA) (to getA)
+instance Definition    ModelKinds where defn     = lensMk defn (error "ambiguous definition (defn) in EquationalRealm") defn
```

Comment:
These errors are just as bad as `undefined`. If the information is not available, or not uniquely available, in all model kinds, then it should be considered available at all.

If it's crucial that the information is available (which could indeed be the case), then this means that the current design is flawed and needs to be revised.

Reply:
I agree. I'm not much of a fan of the errors that I had written for it to be honest.

If we had an InstanceModel/GenDefn/TheoryModel with an EquationalRealm, what would we want to say the;
* "Definition" is? Perhaps we just need a custom "Definition" sentence for an EquationalRealm? I think this sounds about right actually -- I don't know why I didn't think about this before.
* "DefiningExpr" is? Perhaps we need DefiningExpr to be returning a "Maybe Expr" instead of a `Lens c Expr` -- this would be following suit in the current ExprRelat (in that ExprRelat's `relat` function just returns a plain Relation), but against the redesign proposed of `ExprRelat`

Perhaps we should be changing the definition of EquationalRealms (`EquationalRealm QuantityDict (NE.NonEmpty RealmVariant)`) into `EquationalRealm QuantRealm` where a `QuantRealm` contains a `QuantityDict` and `(NE.NonEmpty RealmVariant)`. Then we'd be probably be able to avoid these errors and clean up the code a bit to avoid `undefined`. If the above 2 "Definition" and "DefiningExpr" are actually things we should define instead of creating errors, then I think this would be a decent solution.

Reply:
I actually really liked the idea of having a separate Realm carrier, I think it makes a lot of sense in this context. So, I made a prototype in e45bc854231efe71a8a7e93f45f9376ace92a8b4 (though, it will be easier to review in this branch, since it contains this commit).

It allowed us to define the "Definition" instance, and I think that definition would be good to have as a relational statement between each equation, which may not be something that we want to procedurally generate via the definitions of the underlying variants.

I imagine that if we have only 1 RealmVariant in a Realm, then the DefiningExpr would be that 1 RealmVariant's Expr. However, if we have more than 1, it will be ambiguous. If we change the type of `defnExpr`/DefiningExpr to return a `Maybe (Lens c Expr)`/`Maybe Expr`, then I think we can define something for Realms.

Also, I named the bundle "Realm" but it's probably not the best name for it. Maybe QDefRealm?

Reply:
See the discussion on that commit.  This is getting a lot closer.

However, I definitely don't want to use `Maybe` here. Basically one should not be asking for 'the' definition of something that doesn't have it. So we need to understand why we're even asking the question. So we'll need to see who is using those features of IM/TM/etc and in what way.

I think that in most cases, we'll be able to make sense of what it's trying to do, which will feed back into the design of this feature.

Reply:
I see. I will try to check usage of defnExpr (of DefiningExpr) and relat (of ExprRelat), I think this also relates to the below discussion about `elimMk` too. I'll post when I've understood a bit more about the existing usage.

Reply:
As of right now, `defnExpr` is almost exclusively used on QDefinitions and DataDefinitions. So, it looks like I didn't need (nor should I have) added the extra instance for `DefiningExpr` for IM/TM/GD. `relat` is primarily used for cosmetics and quickly grabbing "x $= ..." Exprs. However, if we intend to replace fully replace `relat` with `defnExpr`, it will require a bit more work since `defnExpr` is not intended for this (and incompatible, because it currently is geared towards uniquely identifying exprs).

If we return a NonEmpty list of Exprs instead of a single Expr, for all the instances that use `relat`, we can just intercalate `$=`, prepending a symbol as well, but then it might not work nicely with the existing usage for DataDefinitions -- for example, defining unique expressions for constants. I think we'll need to have a both in some capacity.

Reply:
Right - this is all very valuable design information that shouldn't be lost as discussion on the PR. Probably it should go in the `Classes` file that defines the various classes (and from there, `relat` and `defnExpr`) as the reasoning behind them.

Once we design what each of them means, then it will be clear what to do. And what to do might include deleting some, creating some new ones (with a different meaning), etc.

Reply:
That's a good idea. I just moved it now. Hopefully it can be helpful in the future. Thank you!